cirq-google/cirq_google/api/v2/ndarrays_test.py

```diff
@@ -0,0 +1,250 @@
+    with pytest.raises(ValueError, match="Cannot convert array"):
+        ndarrays.to_bitarray(data)
+
+
+@pytest.mark.parametrize("dtype", ["u4", "u8"])
+def test_incompatible_int64_dtypes(dtype):
+    array = np.array([1, 2, 3, 4], dtype=dtype)
+    with pytest.raises(ValueError):
+        ndarrays.to_int64_array(array)
+
+
+@pytest.mark.parametrize("dtype", ["u4", "i8"])
+def test_incompatible_int32_dtypes(dtype):
+    array = np.array([1, 2, 3, 4], dtype=dtype)
+    with pytest.raises(ValueError):
+        ndarrays.to_int32_array(array)
+
+
+@pytest.mark.parametrize("dtype", ["i4", "u4", "i8", "u8"])
+def test_incompatible_int16_dtypes(dtype):
+    array = np.array([1, 2, 3, 4], dtype=dtype)
+    with pytest.raises(ValueError):
+        ndarrays.to_int16_array(array)
+
+
+@pytest.mark.parametrize("dtype", ["i1", "u2", "i2"])
+def test_incompatible_uint8_dtypes(dtype):
+    array = np.array([1, 2, 3, 4], dtype=dtype)
+    with pytest.raises(ValueError):
+        ndarrays.to_uint8_array(array)
+
+
+@pytest.mark.parametrize("dtype", ["c16", "<c16", ">c16", "c8"])
+@pytest.mark.parametrize(
+    "values", [[], [1 + 2j, 3 + 4j, 5 + 6j], [[1 + 2j, 2 + 3j], [3 + 4j, 4 + 5j]], [[1, 2], [3, 4]]]
+)
+def test_complex128_conversions(dtype, values):
+    array = np.array(values, dtype=dtype)
+    msg = ndarrays.to_complex128_array(array)
+    assert tuple(msg.shape) == array.shape
+    deserialized = ndarrays.from_complex128_array(msg)
+    assert np.all(array == deserialized)
+    assert array.dtype.byteorder == deserialized.dtype.byteorder
+
+
+@pytest.mark.parametrize("dtype", ["c8", "<c8", ">c8"])
+@pytest.mark.parametrize(
+    "values", [[], [1 + 2j, 3 + 4j, 5 + 6j], [[1 + 2j, 2 + 3j], [3 + 4j, 4 + 5j]], [[1, 2], [3, 4]]]
+)
+def test_complex64_conversions(dtype, values):
+    array = np.array(values, dtype=dtype)
+    msg = ndarrays.to_complex64_array(array)
+    assert tuple(msg.shape) == array.shape
+    deserialized = ndarrays.from_complex64_array(msg)
+    assert np.all(array == deserialized)
+    assert array.dtype.byteorder == deserialized.dtype.byteorder
+
+
+@pytest.mark.parametrize("dtype", ["i2", "f4"])
+def test_incompatible_complex128(dtype):
+    array = np.array([1, 2, 3, 4], dtype=dtype)
+    with pytest.raises(ValueError):
+        ndarrays.to_complex128_array(array)
+
+
+@pytest.mark.parametrize("dtype", ["i2", "f4", "c16", "<c16", ">c16"])
+def test_incompatible_complex64(dtype):
+    array = np.array([1, 2, 3, 4], dtype=dtype)
+    with pytest.raises(ValueError):
+        ndarrays.to_complex64_array(array)
+
+
+def test_overwrite_works():
+    array1 = np.array([1, 2, 3, 4], dtype="i4")
+    array2 = np.array([5, 6], dtype="i4")
+    msg = ndarrays_pb2.Int32Array()
+    ndarrays.to_int32_array(array1, msg)
+    ndarrays.to_int32_array(array2, msg)
+    assert tuple(msg.shape) == array2.shape
+    assert np.all(array2 == ndarrays.from_int32_array(msg))
+
+
+def test_empty_messages():
```

Comment:
for a full coverage, we might add `from_int8_array` and `from_complex64_array` too.